### PR TITLE
[SPARK-45205][SQL] CommandResultExec to override iterator methods to avoid triggering multiple jobs.

### PR DESCRIPTION
**Pull request by yorksity**
### What changes were proposed in this pull request?

After SPARK-35378 was changed, the execution of statements such as ‘show parititions test' became slower. The change point is that the execution process changes from ExecutedCommandEnec to CommandResultExec, but ExecutedCommandExec originally implemented the following method

override def executeToIterator(): Iterator[InternalRow] = sideEffectResult.iterator

CommandResultExec is not rewritten, so when the hasNext method is executed, a job process is created, resulting in increased time-consuming

### Why are the changes needed?

Improve performance when show partitions/tables.

### Does this PR introduce _any_ user-facing change?

No

### How was this patch tested?

Existing tests should cover this.

### Was this patch authored or co-authored using generative AI tooling?

No

### CODE REVIEW

**Comment by beliefer** [comment]
@yorksity Good catch! Could you add the running time of ‘show parititions test' before and after this PR ?

**Comment by beliefer** [comment]
Please make the title of PR more clear.

**Comment by yorksity** [comment]
> Please make the title of PR more clear.

I have modified the title

**Comment by yorksity** [comment]
> @yorksity Good catch! Could you add the running time of ‘show parititions test' before and after this PR ?

I created a table using the following statement and added 200 partitions
create table test(id int, name string) partitioned by (groupid int)
When I executed the show partitions test statement in version 3.1.1, it took about 0.2 seconds, while in version 3.3.1, it took about 0.9 seconds and there were jobs generated,
After rewriting the iterator method on CommandResultExec, the statement takes about 0.2 seconds

**Comment by yorksity** [comment]
@cloud-fan @yaooqinn @viirya @beliefer  

Please review this PR

**Comment by beliefer** [comment]
cc @cloud-fan

File: sql/core/src/main/scala/org/apache/spark/sql/execution/CommandResultExec.scala [hunk 1]
```diff
@@ -40,6 +40,8 @@ case class CommandResultExec(
 
   override def innerChildren: Seq[QueryPlan[_]] = Seq(commandPhysicalPlan)
 
+  override def executeToIterator(): Iterator[InternalRow] = unsafeRows.iterator
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/CommandResultExec.scala:43]: good catch! can we follow `ExecutedCommandExec` and put this method right after `def executeCollect`?

Comment (yorksity) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/CommandResultExec.scala]: Modified, thank you for reviewing.

**Review by beliefer — APPROVED** [review]
LGTM if tests passed.

**Review by viirya — APPROVED** [review]

**Review by yaooqinn — APPROVED** [review]

**Comment by yorksity** [comment]
image
All tests passed.

**Comment by cloud-fan** [comment]
thanks, merging to master/3.5!

**Pull request closed without merging**